### PR TITLE
feat: validator: Load TLS configuration from ConfigMap

### PR DESCRIPTION
**Pull request by akrejcir**
**What this PR does / why we need it**:
This PR moves the TLS configuration of `template-validator` to a ConfigMap. It is mounded as a file in the pod, and validator is able to update its configuration without restarting the pod.

**Which issue(s) this PR fixes**: 
Jira: https://issues.redhat.com/browse/CNV-28716

**Release note**:
```release-note
None
```

### CODE REVIEW

**Comment by codingben** [comment]
@akrejcir Is it simialr to the implementation as in `kubevirt/vm-console-proxy`, e.g. [filewatch.go](https://github.com/kubevirt/vm-console-proxy/blob/main/pkg/filewatch/filewatch.go)? Maybe you can reuse this existing file to from `kubevirt/vm-console-proxy` to avoid code duplication?

**Comment by akrejcir** [comment]
It is exactly the same package. I've mentioned it in the commit message: https://github.com/kubevirt/ssp-operator/pull/1119/commits/3b2e200f14a9ced14b647fdb231e68c5d4487a91

Do you mean that SSP should import `vm-console-proxy`? I would rather avoid it, because it will cause unnecessary indirect dependencies.

File: internal/template-validator/tlsinfo/tlsinfo.go
Comment (0xFelix) [inline internal/template-validator/tlsinfo/tlsinfo.go]: Could be inlined

File: internal/template-validator/tlsinfo/tlsinfo.go
Comment (0xFelix) [inline internal/template-validator/tlsinfo/tlsinfo.go:70]: Inline this?

File: internal/template-validator/tlsinfo/tlsinfo.go
Comment (0xFelix) [inline internal/template-validator/tlsinfo/tlsinfo.go:66]: Is this required to satisfy the linter?

File: internal/template-validator/tlsinfo/tlsinfo.go
Comment (0xFelix) [inline internal/template-validator/tlsinfo/tlsinfo.go:77]: Wouldn't it be enough to only delay it by a second?

File: internal/template-validator/tlsinfo/tlsinfo.go
Comment (0xFelix) [inline internal/template-validator/tlsinfo/tlsinfo.go]: Can you help me to understand what this mutex does?

File: internal/template-validator/validator/app.go
Comment (0xFelix) [inline internal/template-validator/validator/app.go:96]: Inline this?

File: internal/template-validator/tlsinfo/tlsinfo.go
Comment (0xFelix) [inline internal/template-validator/tlsinfo/tlsinfo.go:51]: Inline this?

File: internal/template-validator/tlsinfo/tlsinfo.go
Comment (0xFelix) [inline internal/template-validator/tlsinfo/tlsinfo.go:95]: Consider a const for this err msg fmt?

File: internal/template-validator/tlsinfo/tlsinfo.go
Comment (0xFelix) [inline internal/template-validator/tlsinfo/tlsinfo.go:214]: Inline

File: internal/operands/template-validator/resources.go
Comment (0xFelix) [inline internal/operands/template-validator/resources.go:38]: Use `VirtTemplateValidator` directly?

File: internal/template-validator/tlsinfo/tlsinfo.go
Comment (akrejcir) [reply internal/template-validator/tlsinfo/tlsinfo.go:66]: no. It is to make the `select` non-blocking. It only checks if the `ti.stopWatch` channel is closed, and if not, it continues.

File: internal/template-validator/tlsinfo/tlsinfo.go
Comment (akrejcir) [reply internal/template-validator/tlsinfo/tlsinfo.go:77]: Without the early stop? I'm not sure, IMHO code should unblock once the stop channel is closed.

File: internal/template-validator/tlsinfo/tlsinfo.go
Comment (akrejcir) [reply internal/template-validator/tlsinfo/tlsinfo.go]: It guards access to the fields `ti.cert` and `ti.certError`.
These fields are accessed from 2 goroutines:
- The file watcher goroutine, that updates these fields when the mouned secret is changed.
- The TLS listener goroutine, that reads the certificate when a new incoming connection is accepted.

File: internal/template-validator/tlsinfo/tlsinfo.go
Comment (0xFelix) [reply internal/template-validator/tlsinfo/tlsinfo.go:77]: Yes that's what I was suggesting. Not sure if it makes a huge difference to have the early stop parallel to the delay. The next loop in the goroutine will also begin by checking if the stop was triggered.

File: internal/operands/template-validator/resources.go
Comment (akrejcir) [reply internal/operands/template-validator/resources.go:38]: I wanted to keep the pattern as other constants. I can change it if you want.

File: internal/template-validator/tlsinfo/tlsinfo.go
Comment (akrejcir) [reply internal/template-validator/tlsinfo/tlsinfo.go]: done

File: internal/template-validator/validator/app.go
Comment (akrejcir) [reply internal/template-validator/validator/app.go]: done

File: internal/template-validator/tlsinfo/tlsinfo.go
Comment (akrejcir) [reply internal/template-validator/tlsinfo/tlsinfo.go]: done

File: internal/template-validator/tlsinfo/tlsinfo.go
Comment (akrejcir) [reply internal/template-validator/tlsinfo/tlsinfo.go]: done

File: internal/template-validator/tlsinfo/tlsinfo.go
Comment (0xFelix) [reply internal/template-validator/tlsinfo/tlsinfo.go:69]: ACK

File: internal/template-validator/tlsinfo/tlsinfo.go
Comment (0xFelix) [reply internal/template-validator/tlsinfo/tlsinfo.go]: ACK

File: internal/operands/template-validator/resources.go
Comment (0xFelix) [reply internal/operands/template-validator/resources.go:38]: Let's stay consistent for now. It could be changed in a follow up.

File: internal/template-validator/tlsinfo/tlsinfo.go
Comment (0xFelix) [inline internal/template-validator/tlsinfo/tlsinfo.go:86]: Personal preference, but move it to the beginning of the file?

File: internal/template-validator/tlsinfo/tlsinfo.go
Comment (0xFelix) [inline internal/template-validator/tlsinfo/tlsinfo.go:175]: Is it an error of level Info? Shouldn't it be higher?

Comment (0xFelix) [reply internal/template-validator/tlsinfo/tlsinfo.go:175]: Also, what did it fail to do? :)

File: internal/template-validator/tlsinfo/tlsinfo_test.go
Comment (0xFelix) [inline internal/template-validator/tlsinfo/tlsinfo_test.go:90]: `defer` vs `DeferCleanup`?

File: internal/template-validator/tlsinfo/tlsinfo_test.go
Comment (0xFelix) [inline internal/template-validator/tlsinfo/tlsinfo_test.go:138]: Which error should occur?

File: internal/template-validator/tlsinfo/tlsinfo_test.go
Comment (0xFelix) [inline internal/template-validator/tlsinfo/tlsinfo_test.go:110]: Which error?

File: internal/template-validator/tlsinfo/tlsinfo_test.go
Comment (0xFelix) [inline internal/template-validator/tlsinfo/tlsinfo_test.go:152]: Same as above, and also below

File: internal/template-validator/tlsinfo/tlsinfo_test.go
Comment (akrejcir) [reply internal/template-validator/tlsinfo/tlsinfo_test.go:90]: `defer` is called when `func() {}` returns. `DeferCleanup()` is called after the `It()` test case, similarly to `AfterEach()`.
 It is useful to call in `BeforeEach()` to register cleanups that will be called after the test.

File: internal/template-validator/tlsinfo/tlsinfo.go
Comment (akrejcir) [reply internal/template-validator/tlsinfo/tlsinfo.go:86]: I prefer it to be close the where it is used.

File: internal/template-validator/tlsinfo/tlsinfo.go
Comment (akrejcir) [reply internal/template-validator/tlsinfo/tlsinfo.go]: Yes, changed it to error.

File: internal/template-validator/tlsinfo/tlsinfo.go
Comment (0xFelix) [inline internal/template-validator/tlsinfo/tlsinfo.go:200]: What happens if this leaves us with an empty slice in case of user misconfiguration?

File: internal/template-validator/tlsinfo/tlsinfo_test.go
Comment (0xFelix) [inline internal/template-validator/tlsinfo/tlsinfo_test.go:25]: Can it be moved into the `Describe` below?

File: internal/template-validator/tlsinfo/tlsinfo_test.go
Comment (0xFelix) [reply internal/template-validator/tlsinfo/tlsinfo_test.go:90]: Does it make a difference using `DeferCleanup` here?

File: tests/crypto_policy_test.go
Comment (0xFelix) [inline tests/crypto_policy_test.go:209]: Not sure I understand this change. Why not keep the Expects? Looks more concise to me.

File: internal/template-validator/tlsinfo/tlsinfo.go
Comment (akrejcir) [reply internal/template-validator/tlsinfo/tlsinfo.go:200]: Good point. If user passes unknown cipher names, they will be silently left out. That is how it is implemented in the library function `crypto.OpenSSLToIANACipherSuites()`. I will add code to return an error instead.

File: internal/template-validator/tlsinfo/tlsinfo_test.go
Comment (akrejcir) [reply internal/template-validator/tlsinfo/tlsinfo_test.go:25]: No, because it is used in function `writeCertificate()` that is outside the `Describe`.

File: internal/template-validator/tlsinfo/tlsinfo_test.go
Comment (akrejcir) [reply internal/template-validator/tlsinfo/tlsinfo_test.go:90]: I think only a performance difference.

File: tests/crypto_policy_test.go
Comment (akrejcir) [reply tests/crypto_policy_test.go:209]: Because the next commit calls this function in an `Eventually` loop. The `Expect()` would fail the test immediately.

File: internal/template-validator/tlsinfo/tlsinfo.go
Comment (akrejcir) [reply internal/template-validator/tlsinfo/tlsinfo.go:200]: Thinking about this, it may be desired behavior. Usually, on TCP connection, the server gives client a list of supported ciphers and the client chooses. Any unknown ciphers are simply ignored.
Maybe it is ok to use the same approach for configuration file. User could specify multiple ciphers, and it is not an error if the validator does not know some of them.

What do you think?

File: internal/template-validator/tlsinfo/tlsinfo_test.go
Comment (0xFelix) [reply internal/template-validator/tlsinfo/tlsinfo_test.go:90]: Can it be moved to `AfterEach()`?

File: internal/template-validator/tlsinfo/tlsinfo.go
Comment (0xFelix) [reply internal/template-validator/tlsinfo/tlsinfo.go:200]: I was thinking what happens if you configure the API server with an empty slice if using TLS 1.1/1.2? Will it fallback to default ciphers? I'm not sure we should swallow this error silently. It is misconfiguration of the API server, not a failure in negotiating a cipher during a handshake.

File: internal/template-validator/tlsinfo/tlsinfo_test.go
Comment (akrejcir) [reply internal/template-validator/tlsinfo/tlsinfo_test.go:90]: Yes, but I prefer it this way. It is also how it's used in the validator code `app.go`.

File: internal/template-validator/tlsinfo/tlsinfo.go
Comment (akrejcir) [reply internal/template-validator/tlsinfo/tlsinfo.go:200]: If we configure the server ciphers with empty slice `[]` and not use TLS 1.3, then no ciphers will be enabled, so it will not accept any connection. If we configure it with `nil` slice, then it will use default ciphers.
This is IMHO a bad API design in the standard library. Usually in other places there is no functional difference between `[]` and `nil`.

Ok, I will change the code, so it fails with error if ciphers in config file are not known.

Comment (akrejcir) [reply internal/template-validator/tlsinfo/tlsinfo.go:200]: I've implemented the error checking and noticed that OpenShift's TLS security profiles also contain ciphers that are not implemented in the TLS in golang and they are silently ignored.
For example, here is the TLS `Intermediate` profile:
https://github.com/openshift/api/blob/a2817b89f7e0989016967055a142d5e88bed18de/config/v1/types_tlssecurityprofile.go#L288-L303
```go
TLSProfileIntermediateType: {
	Ciphers: []string{
		"TLS_AES_128_GCM_SHA256",
		"TLS_AES_256_GCM_SHA384",
		"TLS_CHACHA20_POLY1305_SHA256",
		"ECDHE-ECDSA-AES128-GCM-SHA256",
		"ECDHE-RSA-AES128-GCM-SHA256",
		"ECDHE-ECDSA-AES256-GCM-SHA384",
		"ECDHE-RSA-AES256-GCM-SHA384",
		"ECDHE-ECDSA-CHACHA20-POLY1305",
		"ECDHE-RSA-CHACHA20-POLY1305",
		"DHE-RSA-AES128-GCM-SHA256",
		"DHE-RSA-AES256-GCM-SHA384",
	},
	MinTLSVersion: VersionTLS12,
},
```

It contains ciphers `DHE-RSA-AES128-GCM-SHA256` and `DHE-RSA-AES256-GCM-SHA384`, which are not available in the golang TLS library.
The code that does the translation between names and cipher IDs ignores them: https://github.com/openshift/library-go/blob/64780a4acea5ee3e565683ad96b1f4f0799f0bd2/pkg/crypto/crypto.go#L149-L180

So it is not trivial to check if a cipher name is valid. We don't want it to fail if users set one of the OpenShift's profiles, but they already contain unknown ciphers. So we would need to specifically allow them, even if they are not implemented.

What do you think?

Comment (0xFelix) [reply internal/template-validator/tlsinfo/tlsinfo.go:200]: As long as an empty slice that is not nil does not result in default values being used, I'm fine with it.

But strange that these profiles ship ciphers that are not supported. Are they supported in downstream distributions of golang?

Comment (akrejcir) [reply internal/template-validator/tlsinfo/tlsinfo.go:200]: The downstream TLS library is the same as upstream, there are no changes.

File: internal/template-validator/tlsinfo/tlsinfo.go
Comment (0xFelix) [inline internal/template-validator/tlsinfo/tlsinfo.go:233]: What if `ti.cert` is nil? You check `ti.certError` above but it might be worth to avoid a nil-ptr access by adding a second check.

Comment (0xFelix) [reply internal/template-validator/tlsinfo/tlsinfo.go:233]: Same for `cipherSuites == nil`.

File: tests/crypto_policy_test.go
Comment (0xFelix) [inline tests/crypto_policy_test.go:122]: Why do other endpoints not need an eventually?

Comment (akrejcir) [reply tests/crypto_policy_test.go:122]: Because `applyTLSConfig()` waits until the new configuration is read by SSP. SSP needs to go to `Deploying` state and then back to `Deployed`.
We don't have any way to check if template-validator reloaded the configuration.

File: internal/template-validator/tlsinfo/tlsinfo.go
Comment (akrejcir) [reply internal/template-validator/tlsinfo/tlsinfo.go:233]: In my opinion, it is better to panic in that case. If `ti.certError == nil` and `ti.cert == nil` , then it is a programmer's error and an error message is probably useless to the user. It's better that we catch the bug sooner if we see nil-ptr panic.

Comment (0xFelix) [reply internal/template-validator/tlsinfo/tlsinfo.go:233]: Won't this result in a crash loop possibly? A Pod could otherwise appear as not Ready instead, which would look cleaner to me. Not going to block the PR though.

Comment (akrejcir) [reply internal/template-validator/tlsinfo/tlsinfo.go:237]: You are right, I've added the check.

File: internal/template-validator/tlsinfo/tlsinfo.go
Comment (0xFelix) [inline internal/template-validator/tlsinfo/tlsinfo.go:235]: What about `ti.cipherSuites`? Are we ever allowing them to be `nil`?

Comment (akrejcir) [reply internal/template-validator/tlsinfo/tlsinfo.go:235]: Yes, they are set to `nil` if the config file is empty. See line 183.

Comment (akrejcir) [reply internal/template-validator/tlsinfo/tlsinfo.go:235]: I don't understand your question. Can you explain what do you mean?

Comment (0xFelix) [reply internal/template-validator/tlsinfo/tlsinfo.go:235]: If the config file is empty, then TLS is going to be disabled and therefore `CreateTlsConfig` is not going to get called, correct? To me that means that `CreateTlsConfig` should not accept `ti.cipherSuites` to be `nil`.

Comment (akrejcir) [reply internal/template-validator/tlsinfo/tlsinfo.go:235]: No, TLS is always enabled. If the config file is empty, it just means that SSP did not pass any specific configuration to validator, and it should use the default.

**Comment by jcanocan** [comment]
Nothing to comment from my site. The PR looks great. Nice job!
/lgtm

File: tests/crypto_policy_test.go
Comment (codingben) [inline tests/crypto_policy_test.go:220]: nit: You've mentioned `${serviceName}.${serviceNamespace}` but instead do `${serviceName}.${pod.Namespace}`.

File: tests/crypto_policy_test.go
Comment (codingben) [inline tests/crypto_policy_test.go:221]: What is `subpath`? Specifically I mean how `%d%s` looks like?

**Review by codingben — APPROVED** [review]
/lgtm

File: internal/template-validator/tlsinfo/tlsinfo.go
Comment (0xFelix) [reply internal/template-validator/tlsinfo/tlsinfo.go:235]: ACK

**Review by 0xFelix — APPROVED** [review]
/approve

/hold

Putting hold to give chance to address Ben's comment.

File: tests/crypto_policy_test.go
Comment (akrejcir) [reply tests/crypto_policy_test.go:220]: Yes. I think the Service must be in the same namespace as Deployment and Pods. It is possible to point Service to a pod in different namespace, but not with simple selector that we use.
https://kubernetes.io/docs/concepts/services-networking/service/#services-without-selectors

Comment (codingben) [reply tests/crypto_policy_test.go:220]: Okay :+1:

**Pull request merged**
